src/common/src/util/value_encoding/mod.rs

```diff
@@ -15,94 +15,204 @@
+}
+
+fn serialize_naivedate(days: i32, buf: &mut Vec<u8>) {
+    buf.put_i32_le(days);
+}
+
+fn serialize_naivedatetime(secs: i64, nsecs: u32, buf: &mut Vec<u8>) {
+    buf.put_i64_le(secs);
+    buf.put_u32_le(nsecs);
+}
+
+fn serialize_naivetime(secs: u32, nano: u32, buf: &mut Vec<u8>) {
+    buf.put_u32_le(secs);
+    buf.put_u32_le(nano);
+}
+
+fn deserialize_datum<B: Buf>(ty: &DataType, data: &mut B) -> memcomparable::Result<Datum> {
+    Ok(Some(match *ty {
+        DataType::Int16 => ScalarImpl::Int16(data.get_i16_le()),
+        DataType::Int32 => ScalarImpl::Int32(data.get_i32_le()),
+        DataType::Int64 => ScalarImpl::Int64(data.get_i64_le()),
+        DataType::Float32 => ScalarImpl::Float32(OrderedF32::from(data.get_f32_le())),
+        DataType::Float64 => ScalarImpl::Float64(OrderedF64::from(data.get_f64_le())),
+        DataType::Varchar => ScalarImpl::Utf8(deserialize_str(data)?),
+        DataType::Boolean => ScalarImpl::Bool(deserialize_bool(data)?),
+        DataType::Decimal => ScalarImpl::Decimal(deserialize_decimal(data)?),
+        DataType::Interval => ScalarImpl::Interval(deserialize_interval(data)?),
+        DataType::Time => ScalarImpl::NaiveTime(deserialize_naivetime(data)?),
+        DataType::Timestamp => ScalarImpl::NaiveDateTime(deserialize_naivedatetime(data)?),
+        DataType::Timestampz => ScalarImpl::Int64(data.get_i64_le()),
+        DataType::Date => ScalarImpl::NaiveDate(deserialize_naivedate(data)?),
+        _ => {
+            panic!("Type is unable to be deserialized.")
+        }
+    }))
+}
+
+fn deserialize_str<B: Buf>(data: &mut B) -> memcomparable::Result<String> {
+    let len = data.get_u32_le();
+    let mut bytes = vec![0; len as usize];
+    data.copy_to_slice(&mut bytes);
+    Ok(String::from_utf8(bytes)?)
+}
+
+fn deserialize_bool<B: Buf>(data: &mut B) -> memcomparable::Result<bool> {
+    match data.get_u8() {
+        1 => Ok(true),
+        0 => Ok(false),
+        value => Err(memcomparable::Error::InvalidBoolEncoding(value)),
+    }
+}
+
+fn deserialize_interval<B: Buf>(data: &mut B) -> memcomparable::Result<IntervalUnit> {
+    let months = data.get_i32_le();
+    let days = data.get_i32_le();
+    let ms = data.get_i64_le();
+    Ok(IntervalUnit::new(months, days, ms))
+}
+
+fn deserialize_naivetime<B: Buf>(data: &mut B) -> memcomparable::Result<NaiveTimeWrapper> {
+    let secs = data.get_u32_le();
+    let nano = data.get_u32_le();
+    NaiveTimeWrapper::new_with_secs_nano(secs, nano)
+}
+
+fn deserialize_naivedatetime<B: Buf>(data: &mut B) -> memcomparable::Result<NaiveDateTimeWrapper> {
+    let secs = data.get_i64_le();
+    let nsecs = data.get_u32_le();
+    NaiveDateTimeWrapper::new_with_secs_nsecs(secs, nsecs)
+}
+
+fn deserialize_naivedate<B: Buf>(data: &mut B) -> memcomparable::Result<NaiveDateWrapper> {
+    let days = data.get_i32_le();
+    NaiveDateWrapper::new_with_days(days)
+}
+
+fn serialize_decimal(decimal: &Decimal, byte_array: &mut Vec<u8>) {
```

Comment:
Why decimal encoding looks so complex? Can we simply serialize `mantissa` and `scale`?